{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "66512cd7",
   "metadata": {},
   "source": [
    "# Patterns in Cell Cycle Gene Expression \n",
    "\n",
    "The project is an example of \"omics\" approaches that measure everything in a population - genomics for genes, transcriptomics for RNA transcripts, proteomics for proteins and so on. Big data like this are becoming more and more popular in biology as they are information-rich and relatively cheap. They require a lot of data exploration and organisation, and are even more useful when many big datasets can be compared at once.\n",
    "\n",
    "> \"The dream of every cell is to become two cells.\" - Francois Jacob\n",
    "\n",
    "Cells grow and divide, in the process replicating their DNA, segregating the resulting chromosomes into 2 daughter cells, along with replicating every other part of the cell and checking that all is well at every step. This \"cell cycle\" involves co-ordinated processes including \"gene expression\" - transcribing new RNA, translating new proteins to supply the machinery for each stage, and degrading these when they're not needed.\n",
    "\n",
    "One route to understanding the cell cycle is to measure everything that's going on in cells, and exploring the data to extract patterns. This project uses a dataset of cell-cycle-dependent gene expression in budding yeast *Saccharomyces cerevisiae*, a unicellular eukaryote that is a useful model organism for understanding shared biological processes.\n",
    "\n",
    "The data are taken from the paper:\n",
    "\n",
    "> Investigating Conservation of the Cell-Cycle-Regulated Transcriptional Program in the Fungal Pathogen, *Cryptococcus neoformans*.\n",
    "> Christina M. Kelliher, Adam R. Leman, Crystal S. Sierra, Steven B. Haase. PLoS Genetics, 2016. https://doi.org/10.1371/journal.pgen.1006453\n",
    "\n",
    "The paper compared datasets from 2 different species of yeast. Here we will focus only on one of them, *Saccharomyces cerevisiae*.\n",
    "\n",
    "\n",
    "# Yeast gene expression dataset\n",
    "\n",
    "The *S. cerevisiae*. dataset was generated by first arresting the cell cycle in a culture of yeast cells using a yeast pheromone called alpha factor. Then, the pheromone was washed out, which allows the culture to start the cell cycle synchronously. Samples were taken every 10 minutes and RNA extracted, sequenced, and analysed, giving an RNA-seq or trancriptomic dataset. The dataset gives estimates of the relative amount of mRNA expressed from every yeast gene in the cell across several successive cell cycles.\n",
    "\n",
    "These data are in the file `GSE80474_Scerevisiae_normalized.txt`, which was part of the supplementary data for the paper.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee50e577",
   "metadata": {},
   "source": [
    "## Setting up the analysis environment\n",
    "\n",
    "First we need to load our analysis modules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "882b3d03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analysis modules\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy\n",
    "import statsmodels.api as sm\n",
    "np.set_printoptions(precision=5, suppress=True)  # suppress scientific floatation \n",
    "sns.set(color_codes=True)\n",
    "%matplotlib inline\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.max_rows', 500)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d216f7ef-df82-4010-9eb0-e691e55d6c18",
   "metadata": {},
   "source": [
    "## Read in the data\n",
    "\n",
    "Let's read the data into the python session and look at it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "341af084",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_table(\"GSE80474_Scerevisiae_normalized.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eb11b53-efed-479c-be06-c9818c11aee3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69969923-1488-4e95-864e-859abb707965",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a863582-dbf2-41df-91eb-87d2111e09e8",
   "metadata": {},
   "source": [
    "## What is in the data, really?\n",
    "\n",
    "This is a real-life dataset taken from a real-life published paper.\n",
    "We will have to think about what the data mean. \n",
    "\n",
    "To get started, it's always a good idea to read the paper that the data came from. The main part of the paper says:\n",
    "\n",
    "> Total RNA was extracted from yeast cells at each time point (every 5 minutes for S. cerevisiae [...]) and multiplexed for stranded RNA-Sequencing.\n",
    "\n",
    "The supplementary methods of the paper (Supplementary data S1) give more details on how the data were analysed:\n",
    "\n",
    "> Transcript quantification of annotated yeast genes was performed using alignment files output from STAR and Cufflinks2 [4]. Time point samples from the respective yeasts were then normalized together using the CuffNorm feature. The normalized FPKM gene expression outputs (“genes.fpkm_table”) were used in the analyses presented. To avoid fractional and zero values, 1 was added to every FPKM value in each dataset using the R statistical programming environment [5]. Fractions and zeros were found to interfere with the periodicity algorithms that involved log-transformation of data points (data not shown). Normalized gene expression data for each yeast are available in S1 and S2 Tables. Raw RNA-Sequencing data from this manuscript have been submitted to the NCBI Gene Expression Omnibus (GEO; http://www.ncbi.nlm.nih.gov/geo/) under accession number GSE80474.\n",
    "\n",
    "This probably makes little sense if it's your first time seeing RNA-seq.\n",
    "\n",
    "If you want to know more about the background and how RNA-seq works, the RNAseqlopedia is a great resource: https://rnaseq.uoregon.edu. \n",
    "\n",
    "\n",
    "## What you need to know about the data\n",
    "\n",
    "For this project you need to know:\n",
    "\n",
    "- Each column is a time point\n",
    "- Except the first column, that contains gene names or identifiers (more about those below)\n",
    "- Each row is a gene\n",
    "- The numbers are estimates of how much of each mRNA is present in each time point\n",
    "- Precisely, the estimates are in units are Fragments per Kilobase per Million reads (FPKM)\n",
    "- The authors added 1 every FPKM value to avoid later problems from taking log transformations of zero\n",
    "\n",
    "Because of this, we will rename the first column of the data frame to `gene`, which is a better description of the contents of the column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92f8fdcc-0c1a-411d-853b-45427343c2af",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.rename(columns = {'time_points':'gene'})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566ccc6f-9331-4c6d-ae0f-c385a5a529c2",
   "metadata": {},
   "source": [
    "## Why are they talking about log transformation?\n",
    "\n",
    "Log transformation is important in many kinds of data analysis and quantitative thinking. Log transformation turns multiplication and division (which are hard) into addition and subtraction (which are easier). Log transformation is also useful for displaying data, where if some of your data values are 100x as large as your other values, log transformations can make them easier to visualise. [This YouTube video by Rafael Irizarry explains log transformation in biology](https://www.youtube.com/watch?v=3huF0DwxCtU).\n",
    "\n",
    "To see this, let's take a look at the distribution of some of the data.\n",
    "\n",
    "To get a feel for it, we can focus on a single timepoint at 0 minutes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ddf874d-849c-4a66-919d-cf3e0d6044fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "estimates_0min = df[\"0\"]\n",
    "estimates_0min"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e470d93-b409-4f55-b206-5df11c51ab1a",
   "metadata": {},
   "source": [
    "Plotting a histogram of the distribution in the native scale puts many values close to zero, and doesn't look like anything much:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fe3bccd-3a9b-4fe9-8f9f-8ad633edc1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(x = estimates_0min)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67eecc80-c57e-4a19-adeb-a0cc4178bdcb",
   "metadata": {},
   "source": [
    "Plotting a histogram of the distribution on a log10-scale is more helpful:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12283ea4-8480-4365-8d7d-a4161d5b550b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(x = np.log10(estimates_0min))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bfd696c-646e-4ca2-aa89-e2837efbc73d",
   "metadata": {},
   "source": [
    "Here we see that there are many values close to zero, that's the tall bar on the left.\n",
    "Then there's a broad distribution of expression values over many orders of magnitude.\n",
    "That means that there are some mRNAs that are 10000 times as abundant as other mRNAs.\n",
    "\n",
    "You could check if these patterns hold at other timepoints too.\n",
    "\n",
    "The take-home message: **if your plot doesn't look good or your results don't make sense, try a log-transformation**."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfc75c8b-1142-4aba-8c18-3eddfd1737e6",
   "metadata": {},
   "source": [
    "## Tidying the data\n",
    "\n",
    "These data are not tidy...\n",
    "\n",
    "> In tidy data: Every column is a variable. Every row is an observation.\n",
    "\n",
    "Tidying the data will make it easier to plot.\n",
    "\n",
    "My goal here is that each observation is: one gene, one timepoint, one mRNA abundance. I'm going to call the mRNA abundance estimate FPKM1 to remind me that it is in units of FPKM plus one.\n",
    "\n",
    "So I'll use `df.melt` to make a data frame with columns `gene`, `time_point`, `FPKM1`.\n",
    "We also need to make sure that `time_point` is recorded as a number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b71d29c3-e378-4d92-a96b-bd5ee3f0040c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tidy = df.melt(id_vars = \"gene\", var_name = \"time_point\", value_name = \"FPKM1\")\n",
    "df_tidy[\"time_point\"] = pd.to_numeric(df_tidy[\"time_point\"], downcast=\"integer\")\n",
    "df_tidy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a136a27-c155-43a1-82cd-68a07cc5178e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check the data types\n",
    "df_tidy.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80ddd8f4-1092-4213-8ef4-de6ab21493d4",
   "metadata": {},
   "source": [
    "## Plot abundance for a gene\n",
    "\n",
    "Some genes called \"cyclins\" help to regulate the cell cycle. In yeast these genes are sensibly called CLN1, CLN2, CLN3. There are also B-type cyclins called CLB1 through CLB6.\n",
    "\n",
    "Let's visualise the mRNA abundance for some of these genes across the cell cycle, starting with cyclin 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1746b4c2-563d-4739-88e9-b9361332822b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tidy_CLN1 = df_tidy[df_tidy.gene == \"CLN1\"]\n",
    "df_tidy_CLN1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8ae6905-60eb-41f8-84c7-7476952a43c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.lineplot(data = df_tidy_CLN1, x=\"time_point\", y=\"FPKM1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c37e445c-02fb-48e0-9bbe-b952c8405b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.lineplot(data = df_tidy[df_tidy.gene == \"CLN2\"], x=\"time_point\", y=\"FPKM1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f082b26-ee65-4f19-9847-c1522a497738",
   "metadata": {},
   "source": [
    "Instead of plotting each gene independently, we can plot multiple genes on the same axes in different colours to distinguish them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cebe1ecb-0998-4768-84ba-9ea7186da4b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tidy_CLN123 = df_tidy[df_tidy.gene.isin((\"CLN1\", \"CLN2\", \"CLN3\"))]\n",
    "\n",
    "sns.lineplot(data = df_tidy_CLN123, x=\"time_point\", y=\"FPKM1\", hue = \"gene\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e253b869-41df-4fcf-9f1a-e96cf1ba3940",
   "metadata": {},
   "source": [
    "What if we include both the cyclins and the B-type cyclins?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b5f08ad-c7fe-4937-9dca-1f6216231808",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tidy_CLNCLB = df_tidy[df_tidy.gene.isin((\"CLN1\", \"CLN2\", \"CLN3\", \"CLB1\", \"CLB2\", \"CLB3\", \"CLB4\", \"CLB5\", \"CLB6\"))]\n",
    "\n",
    "sns.lineplot(data = df_tidy_CLNCLB, x=\"time_point\", y=\"FPKM1\", hue = \"gene\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53530805-8020-4bd3-be33-70825d7a0030",
   "metadata": {},
   "source": [
    "That lineplot is incomprehensible, there's too many overlapping colours. We'll have to try a different approach. Let's make a heatmap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a001eb0-2ccc-4d62-aacc-5a4b5ca72eb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tidy_CLNCLB_pivot = df_tidy_CLNCLB.pivot(columns = \"time_point\", index = \"gene\")\n",
    "\n",
    "sns.heatmap(data = df_tidy_CLNCLB_pivot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b8b9e2b-4ff7-46c8-bde0-34951ccac208",
   "metadata": {},
   "source": [
    "In this heatmap it's reasonably clear that CLN3 is expressed first. Then CLN2, CLN1, CLB6 at about the same time. Then CLB1 and CLB2 later. But CLB3 and CLB4 are barely visible because they have lower peak abundance.\n",
    "\n",
    "So let's try to plot on a log scale."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20a0d9de-1c94-4dd7-94fa-954b21bfaa2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.heatmap(data = np.log10(df_tidy_CLNCLB_pivot))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c3c89b6-8020-439d-a6f8-6ba540efca77",
   "metadata": {},
   "source": [
    "That doesn't look better. Let's try once more changing the colourmap to a \"diverging\" colourmap so we can see differences in a different way. \n",
    "\n",
    "[Help page on matplotlib colourmaps](https://matplotlib.org/stable/users/explain/colors/colormaps.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad0795f8-41d4-4dff-b846-68329e99aa26",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.heatmap(data = np.log10(df_tidy_CLNCLB_pivot), cmap=\"bwr\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b51a04e2-2a3b-44cd-8283-6430dbd30b41",
   "metadata": {},
   "source": [
    "This heatmap makes the differences a bit clearer.\n",
    "\n",
    "What else could you do to make the meaning of the figure more clear? What labels and information are needed to tell a story? For a start, which x-axis and y-axis labels would help?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d2871bf-20d3-4672-95fa-fd99c62f69e5",
   "metadata": {},
   "source": [
    "## Log fold-change\n",
    "\n",
    "One problem with these plots is that they show abundance of each mRNA relevant to other mRNAs.\n",
    "\n",
    "What if we need to compare *relative* abundance - i.e. the change in abundance of an mRNA relative to its average? This is an example of scaling and normalisation.\n",
    "\n",
    "A standard and sensible approach is log fold-change:\n",
    "\n",
    "- calculate the log of each FPKM1\n",
    "- calculate the mean log value separately for each gene\n",
    "- the log fold-change is the log value minus the mean value, separately for each gene.\n",
    "\n",
    "You should be able to calculate the log fold-change in a new data frame using tools you've learned on the course, including [pandas.DataFrame.groupby](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html). Mostly in genomics people would use a log2 fold-change because it's easy for humans to interpret 2-fold up or 2-fold down. If you need to remind yourself what log2 (logarithm base 2) is, go ahead, it's important.\n",
    "\n",
    "Then it's helpful to use a heatmap with a diverging colourmap like `bwr`, and with the `center` value set to 0, meaning that when log2 fold change is 0 the gene has its average value. Why not try it?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b660f1e-a8ec-4574-83c9-0e684be6326c",
   "metadata": {},
   "source": [
    "## What do these gene expression plots over time mean?\n",
    "\n",
    "These plots show that different cyclins have high expression at different stages of the cell cycle.\n",
    "\n",
    "Compare that with [Figure 1 of the paper that measures the proportion of cells budded at each timepoint](https://doi.org/10.1371/journal.pgen.1006453.g001).\n",
    "\n",
    "Can you see how the up-and down correspond to the population of cells starting synchronously, and then becoming less synchronous? So that the time of first budding is about 60 minutes in? Then about 140 minutes cells start to produce second buds, but the peak is broader and wider so the times at which budding happens are spread out over the population?\n",
    "\n",
    "This also shows in the gene expression. Cell-cycle dependent genes have tight tall peaks in expression during the first cell cycle, then broader and wider peaks at the cell cycle, and at later cycles the cells are no longer synchronised so the peak is barely visible.\n",
    "\n",
    "When is each cyclin expressed relative to the emergence of buds?\n",
    "\n",
    "Several research project questions could go along similar lines: which other genes are expressed at which point of the cell cycle? Which functions are needed at different points?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2baa8170-6eec-43a6-94e5-7c04b81c8651",
   "metadata": {},
   "source": [
    "## Suggested hypotheses for a project\n",
    "\n",
    "This is a big dataset! There are many potential hypotheses that you can explore using this dataset, or even by comparing it with other datasets.\n",
    "\n",
    "- DNA replication machinery is periodically expressed in the cell cycle.\n",
    "- Cell wall components are periodically expressed in the cell cycle.\n",
    "- Functionally related genes are expressed at the same time as DNA replication machinery.\n",
    "- Distinct cell-cycle promoting cyclins are expressed at different points of the cell cycle.\n",
    "- Ribosomes are made at a distinct point in the cell cycle.\n",
    "- Ribosomal biogenesis genes, that make ribosomal RNA, are made before ribosomal proteins are made.\n",
    "- Gene expression patterns are essentially the same in the first and second cell cycle following arrest.\n",
    "- Cell cycles proceed essentially the same way however we synchronise the cells, if we compare with datasets using other synchronisation methods.\n",
    "- Cell cycle dependent gene expression patterns are evolutionarily conserved, if we compare with datasets from other species.\n",
    "\n",
    "Some of these might use clustering methods such as those we have developed in the course.\n",
    "\n",
    "A very brief [Overview of the Cell Cycle textbook chapter explains the need for DNA replication](https://www.ncbi.nlm.nih.gov/books/NBK26869/)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd3b92d8",
   "metadata": {},
   "source": [
    "# Yeast gene name descriptions\n",
    "\n",
    "To make more sense of this gene expression data, we need to know more about the genes we are measuring. Which of those are involved in DNA replication, for example? We have matched data containing information on protein-coding genes, including brief descriptions about what's known of gene function. These 2 datasets are linked by gene identifiers.\n",
    "\n",
    "The file is `yeast_gene_name_descriptions_2023-09-11.tsv`. \n",
    " \n",
    "It was downloaded from the Saccharomyces Genome Database https://www.yeastgenome.org - this is a huge site that summarises everything known about this yeast. If you want to know more about some yeast genes, you can look up details there and link through to the primary research papers. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e4cedba",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df = pd.read_table(\"yeast_gene_name_descriptions_2023-09-11.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8fccbe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3fcc723-f05a-460f-ad45-ed2975cc7a0a",
   "metadata": {},
   "source": [
    "There is a lot going on here. Each gene has a `primaryIdentifier`, a `secondaryIdentifier`, a `symbol`, and a `name`. Why do they need so many identifiers? One reason is that different databases containing different information need to connect to each other. Another reason is that there are different kinds of biological molecules - for example `YAL001C` is the name of the region of DNA that contains an open reading frame that encodes a protein sequence, and `TFC3`.\n",
    "\n",
    "Let's look again at the gene names that came in our mRNA abundance data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca9b08d5-209a-4421-914f-92512ec21dec",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"gene\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "719ad266-4075-405c-9e35-7a77e98f298f",
   "metadata": {},
   "source": [
    "And compare to the gene name data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12a8e5cd-33a4-4e24-b8ed-33ce9c072141",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df[\"Gene.symbol\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c8e2ccd-c21e-4bcd-a227-94a68016c495",
   "metadata": {},
   "source": [
    "These look different, and even have different numbers of entries... oh no!?!\n",
    "\n",
    "Reminder, this is a real-life example with published data derived from public databases and published papers. The problem of making one set of data correspond with another set of data from a different place is an everyday problem in real-life data. The tools we're working with in this course can help.\n",
    "\n",
    "There are 2 important differences between the genes in these datasets, in fact:\n",
    "1. The mRNA abundance dataset counts non-coding RNAs like `15S_rRNA` as well as messenger RNAs\n",
    "2. The mRNA abundance dataset `gene` column is a mixture of protein names (in the `Gene.symbol` column of `gene_df`) and systematic names (in the `Gene.secondaryIdentifier` column).\n",
    "\n",
    "If you want, you can check this yourself by looking through the lists and searching for some gene names on https://www.yeastgenome.org."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89831e3d-ae1b-48a0-b9fe-29c9f1fbdfed",
   "metadata": {},
   "source": [
    "# Making the datasets talk to each other\n",
    "\n",
    "There are different options as to how to make these agree. Here I will create a new column in `gene_df` called `gene` that contains the `Gene.symbol` if it exists and `Gene.secondaryIdentifier` otherwise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3106a2af-bb5f-43ab-b0c2-d4f9b7709736",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df[\"gene\"] = gene_df[\"Gene.symbol\"].fillna(gene_df[\"Gene.secondaryIdentifier\"])\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93d1c97c-7dd2-4979-badf-5eef55cef3eb",
   "metadata": {},
   "source": [
    "Check that this new column `gene` has the right contents: it's equal to `Gene.symbol` where that's defined, `Gene.secondaryIdentifier` otherwise.\n",
    "\n",
    "Now we can merge the data frames."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b3c4d3-78d6-4259-8eff-a4c491f2d7c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged_all = gene_df.merge(right = df, on = \"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f193cdc-5f46-4f29-94b7-b2f0d04b996a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e55844de-da91-4628-949c-55dc0e8be29c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged_all.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82e40295-f38b-4ecd-9ce9-ef982156b608",
   "metadata": {},
   "source": [
    "Notice that this merged dataFrame has fewer rows than either precursor.\n",
    "\n",
    "I'd guess that this is because some mRNAs weren't detected at all in the expression dataset. Or because the data were generated on a different genome release and the names have changed. It's not always possible to find out without analysing everything again from scratch, and there has to be a really good reason to do that"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18d8623c-62a1-4bc7-a339-8b05d1f5985e",
   "metadata": {},
   "source": [
    "## Exclude dubious genes\n",
    "\n",
    "Let's try focusing on genes where there is good evidence that a functional protein is produced. This is because an Open Reading Frame (ORF) is a predicted gene based on the sequence, but predictions are imperfect. As described in [New data and collaborations at the Saccharomyces Genome Database](https://doi.org/10.1093/genetics/iyab224):\n",
    "\n",
    "> All *S. cerevisiae* ORFs are categorized into one of three groups: “verified” ORFs are those for which there is clear experimental evidence for the presence of expression of a protein-coding gene; “uncharacterized” ORFs are likely, but not yet established, to encode a protein; and “dubious” ORFs are unlikely to encode a protein.\n",
    "\n",
    "If you look at `gene_df` you'll notice that the `Gene.qualifier` column contains values that correspond to these\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0753ce1e-1e36-4442-954d-5607a6054108",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df[\"Gene.qualifier\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44eda7cf-6966-4801-b596-7806d3ae10c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df[\"Gene.qualifier\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "657060ca-236a-4b51-80f9-2ebffd0d66ea",
   "metadata": {},
   "source": [
    "Let's remove the dubious values and see what we have left"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f99855ff-6bf7-4d95-9369-410f2bc97383",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df_nondubious = gene_df[ gene_df[\"Gene.qualifier\"] != \"Dubious\" ]\n",
    "gene_df_nondubious[\"Gene.qualifier\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca569e30-c2a1-4d8e-b96a-f593c5bb441d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df_nondubious.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02fea284-fe92-462b-ae38-83ee1b458273",
   "metadata": {},
   "source": [
    "Now let's merge the nondubious values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "371d1929-070c-44d5-a1ce-c104207fcc7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged = gene_df_nondubious.merge(right = df, on = \"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0423a536-c0f9-48f3-b52d-b29911be5bfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "119de8be-e34e-4e80-8c11-82794ea685bd",
   "metadata": {},
   "source": [
    "Still less ... this is probably the best we can do for now.\n",
    "\n",
    "This merged dataset on non-dubious genes could be a good place to start with connecting functional information in the `gene_df` to the mRNA abundance data. Or, for some analyses it might be easier to first choose a subset of the genes and then analyse the mRNA abundance data accordingly."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59cac015-d9a3-49a4-bb74-b00cdef7a235",
   "metadata": {},
   "source": [
    "# Sanity-saving hint: decide how you want to load and clean up your data\n",
    "\n",
    "This introduction has spent a lot of time loading, inspecting, tidying, and cleaning data.\n",
    "\n",
    "You don't want to do that every time.\n",
    "\n",
    "When you start your project work it's a good idea to:\n",
    "1. decide what format you want the data in\n",
    "2. figure out the minimal code that does these essential steps with your data\n",
    "3. start a new notebook containing the minimal code and test it works\n",
    "\n",
    "This is a sanity-saving approwch. You don't want to have one explosively messy document where you can't tell what's happening. You also don't want a hundred documents that take different approaches so you can't find anything. Try some things, then clean them up, and make a new and cleaner document when it's helpful."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "031b897a-dd58-4445-b0d5-88764286381b",
   "metadata": {},
   "source": [
    "## Picking cyclin-related genes out\n",
    "\n",
    "As one example of how we could proceed to plot a subset of genes, let's find all the genes whose description involves \"cyclin\", then plot those.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2174b9e-bf4a-48b2-aec3-a956a214da5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df_cyclinrelated = gene_df[gene_df[\"Gene.description\"].str.contains(\"cyclin\")]\n",
    "gene_df_cyclinrelated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fced728f-7b94-4cde-aec0-4ae956865652",
   "metadata": {},
   "source": [
    "Wait, that's missing CLN3. Let's search for `cyclin` and `Cyclin` etc, by ignoring the case of the search term.\n",
    "\n",
    "Here I will use a regular expression (`re`) because that's the way I know how to search case-insensitively. I looked up help for [`pandas.Series.str.contains`](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f54f882e-0605-4188-bb9d-3ea184fa2e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "#importing regular expression module\n",
    "import re\n",
    "\n",
    "gene_df_cyclinrelated = gene_df[gene_df[\"Gene.description\"].str.contains(\"cyclin\", flags=re.IGNORECASE)]\n",
    "gene_df_cyclinrelated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c3a57b6-79ad-403d-bcdc-7d77499168a1",
   "metadata": {},
   "source": [
    "We'll use the same idea we used earlier for the CLN and CLB genes, search for  `df_tidy_CLNCLB_pivot` above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79fb42fb-3dcc-4c79-b36a-7c53a73e99a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tidy_cyclinrelated = df_tidy[ df_tidy.gene.isin( gene_df_cyclinrelated.gene )]\n",
    "df_tidy_cyclinrelated_pivot = df_tidy_cyclinrelated.pivot(columns = \"time_point\", index = \"gene\")\n",
    "\n",
    "sns.heatmap(data = np.log10(df_tidy_cyclinrelated_pivot), cmap = \"bwr\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d9ec850-5fcd-40f9-9d06-ad2fafe9aade",
   "metadata": {},
   "source": [
    "There's more work to do here to make an interpretable plot. You could try: \n",
    "- plotting log2 fold-change instead of log10(FPKM1).\n",
    "- giving the axes helpful names and labels.\n",
    "- adjusting plot size so every gene name appears.\n",
    "- plotting fewer genes, after reading the descriptions to make sense of which to choose.\n",
    "- clustering the plot so genes with similar abundance are plotted together.\n",
    "\n",
    "Still, this is an example of how to explore the data."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2a4ec564",
   "metadata": {},
   "source": [
    "# Bud index data\n",
    "\n",
    "We have also shared, courtesy of Christina Kelliher, the data on number of budded cells in S. cerevisiae synchronized with alpha factor, that underlies Figure 1A. This gives you the opportunity to plot the percent budded cells, and compare to the transcriptional data, yourselves.\n",
    "\n",
    "This is found in the excel file `2014-04-21_Kelliher_ScerevisiaeBudCounts.xlsx`, with column names:\n",
    "\n",
    "- Time (min)\n",
    "- 0 Bud\n",
    "- 1 Bud\n",
    "- Total\n",
    "- % 1 bud\n",
    "\n",
    "If you're reading this in to python, you'll need to start with `pandas.read_excel`, and fix the column names to be more python-friendly."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "429c4738-19a5-49d5-88d4-385c1d9bf4ad",
   "metadata": {},
   "source": [
    "# Conclusion\n",
    "\n",
    "Here we presented\n",
    "- a dataset of yeast mRNA abundance during the cell cycle\n",
    "- a dataset of yeast gene names and descriptions\n",
    "- a few ideas about how to approach the data\n",
    "- some potential hypotheses to address using the data\n",
    "\n",
    "Your mission is to think through a hypothesis that you would like to address, then use the data provided and the tools taught in the course to address that hypothesis, and then to explain what you learned."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16ef61f1",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "For background reading, there is an excellent textbook on the cell cycle by David Morgan. In that, Section 3-12 \"Transcriptional Control of Cell-Cycle Regulators\" provides some background. If you are interested, other sections of the book give more information on many other aspects of the cell cycle. (You don't need to read this textbook in order to do the project.)\n",
    "\n",
    "> The cell cycle : principles of control\n",
    "> Morgan, David O.\n",
    "> Northants : Oxford University Press; 2007, ©2007\n",
    "\n",
    "This book is available [in the University library](https://discovered.ed.ac.uk/permalink/44UOE_INST/7g3mt6/alma9913098613502466), and as a [PDF on David Morgan's website](https://morganlab.ucsf.edu/sites/g/files/tkssra2561/f/wysiwyg/Morgan%20Cell%20Cycle%20Book.pdf).\n",
    "\n",
    "David Morgan also has some [online lectures on the cell cycle at iBiology](https://www.ibiology.org/cell-biology/controlling-cell-cycle/).\n",
    "\n",
    "Further references:\n",
    "\n",
    "Comprehensive identification of cell cycle-regulated genes of the yeast Saccharomyces cerevisiae by microarray hybridization. Spellman, P.T. et al. Molecular Biology of the Cell, 1998. https://doi.org/10.1091/mbc.9.12.3273\n",
    "\n",
    "Serial Regulation of Transcriptional Regulators in the Yeast Cell Cycle. Simon, I. et al. Cell, 2001. https://doi.org/10.1016/S0092-8674(01)00494-9\n",
    "\n",
    "Investigating Conservation of the Cell-Cycle-Regulated Transcriptional Program in the Fungal Pathogen, Cryptococcus neoformans. Kelliher, C.M. et al. PLoS Genetics, 2016. https://doi.org/10.1371/journal.pgen.1006453\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1c01b4c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
